File: src/main/java/cspfj/constraint/extension/Matrix2D.java
package cspfj.constraint.extension;

import cspfj.util.BitVector;

public final class Matrix2D implements Matrix {
    private BitVector[] xMatrix;
    private BitVector[] yMatrix;
    private boolean empty;

    public Matrix2D(final int xSize, final int ySize, final boolean initialState) {

        xMatrix = new BitVector[xSize];
        for (int i = xMatrix.length; --i >= 0;) {
            xMatrix[i] = BitVector.factory(ySize, initialState);
        }

        yMatrix = new BitVector[ySize];

        for (int i = yMatrix.length; --i >= 0;) {
            yMatrix[i] = BitVector.factory(xSize, initialState);
        }
        empty = initialState;
    }

    @Override
    public boolean check(final int[] tuple) {
        return xMatrix[tuple[0]].get(tuple[1]);
    }

    @Override
    public void set(final int[] tuple, final boolean status) {
        xMatrix[tuple[0]].set(tuple[1], status);
        yMatrix[tuple[1]].set(tuple[0], status);
        empty = false;
    }

    public BitVector getBitVector(final int position, final int index) {
        switch (position) {
        case 0:
            return xMatrix[index];
        case 1:
            return yMatrix[index];
        default:
            throw new IllegalArgumentException();
        }
    }

    public Matrix2D clone() throws CloneNotSupportedException {
        final Matrix2D matrix2d = (Matrix2D) super.clone();

        matrix2d.xMatrix = new BitVector[xMatrix.length];

        for (int j = xMatrix.length; --j >= 0;) {
            matrix2d.xMatrix[j] = xMatrix[j].clone();
        }

        matrix2d.yMatrix = new BitVector[yMatrix.length];
        for (int j = yMatrix.length; --j >= 0;) {
            matrix2d.yMatrix[j] = yMatrix[j].clone();
        }

        return matrix2d;

    }

    public String toString() {
        final StringBuilder stb = new StringBuilder();
        final int[] tuple = new int[2];
        for (int i = 0; i < xMatrix.length; i++) {
            tuple[0] = i;
            for (int j = 0; j < yMatrix.length; j++) {
                tuple[1] = j;
                if (check(tuple)) {
                    stb.append(1);
                } else {
                    stb.append(0);
                }
            }
            stb.append('\n');
        }
        return stb.toString();
    }

    public boolean isEmpty() {
        return empty;
    }

}

File: src/main/scala/cspfj/constraint/extension/Matrix2D.scala
package cspfj.constraint.extension;

import cspfj.util.BitVector;

final class Matrix2D(xSize: Int, ySize: Int, initialState: Boolean) extends Matrix {
  private var xMatrix =
    new Array[BitVector](xSize).map(_ => {
      val bv = BitVector.newBitVector(ySize)
      if (initialState) bv.fill(true)
      bv
    })

  private var yMatrix =
    new Array[BitVector](ySize).map(_ => {
      val bv = BitVector.newBitVector(xSize)
      if (initialState) bv.fill(true)
      bv
    })

  private var empty = initialState

  override def check(tuple: Array[Int]) = xMatrix(tuple(0)).get(tuple(1))

  override def set(tuple: Array[Int], status: Boolean) {
    xMatrix(tuple(0)).set(tuple(1), status);
    yMatrix(tuple(1)).set(tuple(0), status);
    empty = false;
  }

  def getBitVector(position: Int, index: Int) = position match {
    case 0 => xMatrix(index);
    case 1 => yMatrix(index);
    case _ => throw new IllegalArgumentException;
  }

  def copy = {
    val matrix2d = new Matrix2D(xSize, ySize, initialState)
    matrix2d.xMatrix = xMatrix map (_.clone)
    matrix2d.yMatrix = yMatrix map (_.clone)
    matrix2d;
  }

  override def toString = {
    val stb = new StringBuilder;
    for (i <- 0 until xMatrix.length) {
      for (j <- 0 until yMatrix.length) {
        if (check(Array(i, j))) {
          stb.append(1)
        } else {
          stb.append(0)
        }
      }
      stb.append('\n')
    }
    stb.toString
  }

  def isEmpty = empty

}
